### ISSUE TITLE
ansible serial do not  continue when one host failed  in v2.1

### ISSUE THREAD

**Issue by ghost** [ansible/ansible#16241]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
V2.1.0.0
```
##### OS / ENVIRONMENT

Centos 6.5
##### SUMMARY

In the 2.1 version, set serial to 10, if one host fails, then the next round will not execute after execution.
This problem does not exist in 2.0.1.0
##### STEPS TO REPRODUCE

```

```

host has 10 hosts:
172.16.105.1
172.16.105.2
172.16.105.3
172.16.105.4
172.16.105.5
172.16.105.6
172.16.105.7
172.16.105.8
172.16.105.9

site.yml:
- hosts: "*"
  remote_user: admin
  serial: "30%"
##### ACTUAL RESULTS

```
PLAY RECAP *********************************************************************
172.16.105.1              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.2              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.3              : ok=2    changed=0    unreachable=0    failed=1   

playbook would break when one hosts failed!!

In v2.0.1.0
PLAY RECAP *********************************************************************
172.16.105.1              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.2              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.3              : ok=2    changed=0    unreachable=0    failed=1   
172.16.105.4              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.5              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.6              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.7              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.8              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.9              : ok=4    changed=2    unreachable=0    failed=0   
172.16.105.10              : ok=4    changed=2    unreachable=0    failed=0   
```

**Comment by ghost** [comment]
# playbook_executor.py

for batch in self._get_serialized_batches(new_play):
                            if len(batch) == 0:
                                self._tqm.send_callback('v2_playbook_on_play_start', new_play)
                                self._tqm.send_callback('v2_playbook_on_no_hosts_matched')
                                break
                            # restrict the inventory to the hosts in the serialized batch
                            self._inventory.restrict_to_hosts(batch)
                            # and run it...
                            result = self._tqm.run(play=play)

```
                        # break the play if the result equals the special return code
                        if result == self._tqm.RUN_FAILED_BREAK_PLAY:
                            result = self._tqm.RUN_FAILED_HOSTS
                            break_play = True

                        # check the number of failures here, to see if they're above the maximum
                        # failure percentage allowed, or if any errors are fatal. If either of those
                        # conditions are met, we break out, otherwise we only break out if the entire
                        # batch failed
                        failed_hosts_count = len(self._tqm._failed_hosts) + len(self._tqm._unreachable_hosts)
                        if new_play.max_fail_percentage is not None and \
                           int((new_play.max_fail_percentage)/100.0 * len(batch)) > int((len(batch) - failed_hosts_count) / len(batch) * 100.0):
                            break_play = True
                            break
                        elif len(batch) == failed_hosts_count:
                            break_play = True
                            break

                        # save the unreachable hosts from this batch
                        self._unreachable_hosts.update(self._tqm._unreachable_hosts)

                        # if the last result wasn't zero or 3 (some hosts were unreachable),
                        # break out of the serial batch loop
                        if result not in (self._tqm.RUN_OK, self._tqm.RUN_UNREACHABLE_HOSTS):
                            break
```

when one host failed ,the result is RUN_FAILED_HOSTS,  and the play would break. So the max_fail_percentage makes no sense

**Comment by mverwijs** [comment]
I run into this problem without using the 'serial' keyword.

Inventory file:

```
[monkeys]
10.152.33.10

[zebras]
10.152.33.11

[kangaroos]
192.168.76.90
192.168.77.10

```

Playbook

```
- hosts: all

- hosts: monkeys
  tasks:
    - debug: msg="I am a monkey"

- hosts: zebras
  tasks:
    - debug: msg="I am a zebra"

- hosts: kangaroos
  tasks:
    - debug: msg="I am a kanga"
```

Ansible (2.1) command:

```
ansible-playbook -i testinventory test.yml
```

Expected result : all Plays are executed.

Actual result: Only the first play is executed.

```
> ansible-playbook -i testinventory test.yml 

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
fatal: [192.168.76.90]: UNREACHABLE! => {"changed": false, "msg": "SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh", "unreachable": true}
ok: [192.168.77.10]
ok: [10.152.33.10]
ok: [10.152.33.11]

PLAY [monkeys] *****************************************************************

TASK [setup] *******************************************************************
ok: [10.152.33.10]

TASK [debug] *******************************************************************
ok: [10.152.33.10] => {
    "msg": "I am a monkey"
}

PLAY RECAP *********************************************************************
10.152.33.10               : ok=3    changed=0    unreachable=0    failed=0   
10.152.33.11               : ok=1    changed=0    unreachable=0    failed=0   
192.168.76.90              : ok=0    changed=0    unreachable=1    failed=0   
192.168.77.10              : ok=1    changed=0    unreachable=0    failed=0   
```
- when using ansible 1.9.x or 2.0.x, the bug is not triggered and all Plays get executed.
- when running `ansible-playbook -i testinventory test.yml --list-tasks`, all Plays get displayed.
- when /more/ than 1 (one) host is unreachable, the bug is /not/ triggered.
- when removing the `-hosts: all` line from `test.yml`, the bug is /not/ triggered.

Ansible was installed on Ubuntu 14.04:
- apt-get install python-virtualenv
- virtualenv ansible2.1
- . ansible2.1/bin/activate
- pip install ansible=2.1

**Comment by wysekm** [comment]
Hasn't it always been a feature? https://github.com/ansible/ansible/issues/1663

**Comment by tomasz-c** [comment]
> Hasn't it always been a feature? #1663

No. It has not been implemented.
In previous versions, this error did not occur.

Is it possible to force it to run all hosts?

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

This was fixed in 2.1.2.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
